File: memory.cpp
// memory.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <random>
#include <chrono>
#include <thread>
#include <fstream>
#include <json/json.h> // sudo apt-get install libjsoncpp-dev

using namespace std;

const string RESET = "\033[0m";
const string RED = "\033[91m";
const string GREEN = "\033[92m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string MAGENTA = "\033[95m";
const string CYAN = "\033[96m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

vector<string> symbols = {"A","B","C","D","E","F","G","H","I","J","K","L","M",
                          "N","O","P","Q","R","S","T","U","V","W","X","Y","Z",
                          "1","2","3","4","5","6","7","8","9"};
vector<string> symbolColors = {RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN};

string getSymbolColor(const string& sym) {
    return symbolColors[hash<string>{}(sym) % symbolColors.size()];
}

string getHomeDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home);
}

string getRecordFile() {
    return getHomeDir() + "/.memory_records.json";
}

Json::Value loadRecords() {
    ifstream f(getRecordFile());
    Json::Value root;
    if (!f) return root;
    f >> root;
    return root;
}

void saveRecords(const Json::Value& records) {
    ofstream f(getRecordFile());
    f << records.toStyledString();
}

vector<vector<string>> generateBoard(int size) {
    int numPairs = (size * size) / 2;
    vector<string> selected(symbols.begin(), symbols.begin() + numPairs);
    vector<string> board;
    for (string s : selected) {
        board.push_back(s);
        board.push_back(s);
    }
    random_device rd;
    mt19937 g(rd());
    shuffle(board.begin(), board.end(), g);
    vector<vector<string>> result(size, vector<string>(size));
    for (int i = 0; i < size; ++i)
        for (int j = 0; j < size; ++j)
            result[i][j] = board[i * size + j];
    return result;
}

void displayBoard(const vector<vector<string>>& board, const vector<vector<bool>>& revealed, int size) {
    cout << colorize("\n  ", BOLD);
    for (int j = 0; j < size; ++j) cout << colorize("   " + to_string(j+1), BOLD);
    cout << endl;
    for (int i = 0; i < size; ++i) {
        cout << colorize(to_string(i+1) + " ", BOLD);
        for (int j = 0; j < size; ++j) {
            if (revealed[i][j]) {
                string sym = board[i][j];
                string col = getSymbolColor(sym);
                cout << colorize(" " + sym + " ", col);
            } else {
                cout << colorize(" ■ ", BOLD);
            }
        }
        cout << endl;
    }
}

void clearScreen() {
    cout << "\033[2J\033[1;1H";
}

pair<int,int> getCoordinates(int size) {
    while (true) {
        string input;
        cout << "Выберите карточку (строка столбец) или 'q' для выхода: ";
        cin >> input;
        if (input == "q" || input == "Q") {
            cout << colorize("Выход из игры.", YELLOW) << endl;
            exit(0);
        }
        int row, col;
        if (sscanf(input.c_str(), "%d %d", &row, &col) == 2) {
            if (row >= 1 && row <= size && col >= 1 && col <= size) {
                return {row-1, col-1};
            }
        }
        cout << colorize("Неверный ввод. Введите два числа через пробел.", RED) << endl;
    }
}

int main(int argc, char* argv[]) {
    int size = 4;
    int timeout = 0;
    string level;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "-s" && i+1 < argc) size = stoi(argv[++i]);
        else if (arg == "-l" && i+1 < argc) {
            level = argv[++i];
            if (level == "easy") size = 4;
            else if (level == "medium") size = 6;
            else if (level == "hard") size = 8;
        }
        else if (arg == "-t" && i+1 < argc) timeout = stoi(argv[++i]);
        else if (arg == "-h") {
            cout << "Usage: memory [options]\n  -s <N>    Size NxN\n  -l <level> easy|medium|hard\n  -t <sec>  Timeout per move\n";
            return 0;
        }
    }
    if (size % 2 != 0) {
        cout << colorize("Размер должен быть чётным.", RED) << endl;
        return 1;
    }

    Json::Value records = loadRecords();
    string key = to_string(size);

    auto board = generateBoard(size);
    vector<vector<bool>> revealed(size, vector<bool>(size, false));
    int moves = 0;
    int pairsFound = 0;
    int totalPairs = (size * size) / 2;
    auto start = chrono::steady_clock::now();

    while (pairsFound < totalPairs) {
        clearScreen();
        cout << colorize("🧠  МЕМОРИ  |  Размер " + to_string(size) + "×" + to_string(size) +
                         "  |  Ходы: " + to_string(moves) + "  |  Пары: " + to_string(pairsFound) + "/" + to_string(totalPairs), BOLD) << endl;
        displayBoard(board, revealed, size);

        // Первая карточка
        auto [r1, c1] = getCoordinates(size);
        if (revealed[r1][c1]) {
            cout << colorize("Карточка уже открыта.", YELLOW) << endl;
            continue;
        }
        revealed[r1][c1] = true;
        clearScreen();
        cout << colorize("🧠  МЕМОРИ  |  Размер " + to_string(size) + "×" + to_string(size) +
                         "  |  Ходы: " + to_string(moves) + "  |  Пары: " + to_string(pairsFound) + "/" + to_string(totalPairs), BOLD) << endl;
        displayBoard(board, revealed, size);

        // Вторая карточка
        auto [r2, c2] = getCoordinates(size);
        if ((r2 == r1 && c2 == c1) || revealed[r2][c2]) {
            cout << colorize("Неверный выбор.", YELLOW) << endl;
            revealed[r1][c1] = false;
            continue;
        }
        revealed[r2][c2] = true;
        moves++;
        clearScreen();
        cout << colorize("🧠  МЕМОРИ  |  Размер " + to_string(size) + "×" + to_string(size) +
                         "  |  Ходы: " + to_string(moves) + "  |  Пары: " + to_string(pairsFound) + "/" + to_string(totalPairs), BOLD) << endl;
        displayBoard(board, revealed, size);

        if (board[r1][c1] == board[r2][c2]) {
            cout << colorize("✅ Пара найдена!", GREEN) << endl;
            pairsFound++;
            this_thread::sleep_for(chrono::seconds(1));
        } else {
            cout << colorize("❌ Не совпало.", RED) << endl;
            this_thread::sleep_for(chrono::seconds(timeout > 0 ? timeout : 1));
            revealed[r1][c1] = false;
            revealed[r2][c2] = false;
        }
    }

    auto end = chrono::steady_clock::now();
    int elapsed = chrono::duration_cast<chrono::seconds>(end - start).count();
    cout << colorize("\n🎉 Поздравляем! Вы завершили игру за " + to_string(moves) + " ходов и " + to_string(elapsed) + " секунд.", GREEN) << endl;

    // Обновление рекорда
    Json::Value best = records[key];
    if (best.isNull() || moves < best["moves"].asInt() || (moves == best["moves"].asInt() && elapsed < best["time"].asInt())) {
        best["moves"] = moves;
        best["time"] = elapsed;
        time_t now = time(nullptr);
        char buf[64];
        strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", localtime(&now));
        best["date"] = string(buf);
        records[key] = best;
        saveRecords(records);
        cout << colorize("🏆 Новый рекорд для размера " + to_string(size) + "×" + to_string(size) + "!", YELLOW) << endl;
    } else {
        cout << colorize("Лучший результат для этого размера: " + to_string(best["moves"].asInt()) + " ходов за " + to_string(best["time"].asInt()) + " сек.", BLUE) << endl;
    }
    return 0;
}

File: memory.go
// memory.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const (
	reset   = "\033[0m"
	red     = "\033[91m"
	green   = "\033[92m"
	yellow  = "\033[93m"
	blue    = "\033[94m"
	magenta = "\033[95m"
	cyan    = "\033[96m"
	bold    = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var symbols = []string{"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
	"N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z",
	"1", "2", "3", "4", "5", "6", "7", "8", "9"}
var symbolColors = []string{red, green, yellow, blue, magenta, cyan}

func getSymbolColor(sym string) string {
	return symbolColors[hash(sym)%len(symbolColors)]
}

func hash(s string) int {
	h := 0
	for _, c := range s {
		h = h*31 + int(c)
	}
	return h
}

type Record struct {
	Moves int    `json:"moves"`
	Time  int    `json:"time"`
	Date  string `json:"date"`
}

func getHomeDir() string {
	home, _ := os.UserHomeDir()
	return home
}

func getRecordFile() string {
	return getHomeDir() + "/.memory_records.json"
}

func loadRecords() map[string]Record {
	data, err := os.ReadFile(getRecordFile())
	if err != nil {
		return make(map[string]Record)
	}
	var records map[string]Record
	json.Unmarshal(data, &records)
	return records
}

func saveRecords(records map[string]Record) {
	data, _ := json.MarshalIndent(records, "", "  ")
	os.WriteFile(getRecordFile(), data, 0644)
}

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func generateBoard(size int) [][]string {
	numPairs := (size * size) / 2
	selected := symbols[:numPairs]
	var board []string
	for _, s := range selected {
		board = append(board, s, s)
	}
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(board), func(i, j int) { board[i], board[j] = board[j], board[i] })
	result := make([][]string, size)
	for i := range result {
		result[i] = make([]string, size)
		for j := range result[i] {
			result[i][j] = board[i*size+j]
		}
	}
	return result
}

func displayBoard(board [][]string, revealed [][]bool, size int) {
	fmt.Print(colorize("\n  ", bold))
	for j := 0; j < size; j++ {
		fmt.Print(colorize(fmt.Sprintf("   %d", j+1), bold))
	}
	fmt.Println()
	for i := 0; i < size; i++ {
		fmt.Print(colorize(fmt.Sprintf("%d ", i+1), bold))
		for j := 0; j < size; j++ {
			if revealed[i][j] {
				sym := board[i][j]
				col := getSymbolColor(sym)
				fmt.Print(colorize(" "+sym+" ", col))
			} else {
				fmt.Print(colorize(" ■ ", bold))
			}
		}
		fmt.Println()
	}
}

func getCoordinates(size int) (int, int) {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print("Выберите карточку (строка столбец) или 'q' для выхода: ")
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)
		if input == "q" || input == "Q" {
			fmt.Println(colorize("Выход из игры.", yellow))
			os.Exit(0)
		}
		parts := strings.Fields(input)
		if len(parts) == 2 {
			row, err1 := strconv.Atoi(parts[0])
			col, err2 := strconv.Atoi(parts[1])
			if err1 == nil && err2 == nil && row >= 1 && row <= size && col >= 1 && col <= size {
				return row - 1, col - 1
			}
		}
		fmt.Println(colorize("Неверный ввод. Введите два числа через пробел.", red))
	}
}

func main() {
	size := 4
	timeout := 0
	level := ""

	for i := 1; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-s":
			if i+1 < len(os.Args) {
				size, _ = strconv.Atoi(os.Args[i+1])
				i++
			}
		case "-l":
			if i+1 < len(os.Args) {
				level = os.Args[i+1]
				switch level {
				case "easy":
					size = 4
				case "medium":
					size = 6
				case "hard":
					size = 8
				}
				i++
			}
		case "-t":
			if i+1 < len(os.Args) {
				timeout, _ = strconv.Atoi(os.Args[i+1])
				i++
			}
		case "-h":
			fmt.Println("Usage: memory [options]\n  -s <N>    Size NxN\n  -l <level> easy|medium|hard\n  -t <sec>  Timeout per move")
			return
		}
	}
	if size%2 != 0 {
		fmt.Println(colorize("Размер должен быть чётным.", red))
		return
	}

	records := loadRecords()
	key := strconv.Itoa(size)

	board := generateBoard(size)
	revealed := make([][]bool, size)
	for i := range revealed {
		revealed[i] = make([]bool, size)
	}
	moves := 0
	pairsFound := 0
	totalPairs := (size * size) / 2
	start := time.Now()

	for pairsFound < totalPairs {
		clearScreen()
		fmt.Println(colorize(fmt.Sprintf("🧠  МЕМОРИ  |  Размер %d×%d  |  Ходы: %d  |  Пары: %d/%d",
			size, size, moves, pairsFound, totalPairs), bold))
		displayBoard(board, revealed, size)

		// Первая карточка
		r1, c1 := getCoordinates(size)
		if revealed[r1][c1] {
			fmt.Println(colorize("Карточка уже открыта.", yellow))
			continue
		}
		revealed[r1][c1] = true
		clearScreen()
		fmt.Println(colorize(fmt.Sprintf("🧠  МЕМОРИ  |  Размер %d×%d  |  Ходы: %d  |  Пары: %d/%d",
			size, size, moves, pairsFound, totalPairs), bold))
		displayBoard(board, revealed, size)

		// Вторая карточка
		r2, c2 := getCoordinates(size)
		if (r2 == r1 && c2 == c1) || revealed[r2][c2] {
			fmt.Println(colorize("Неверный выбор.", yellow))
			revealed[r1][c1] = false
			continue
		}
		revealed[r2][c2] = true
		moves++
		clearScreen()
		fmt.Println(colorize(fmt.Sprintf("🧠  МЕМОРИ  |  Размер %d×%d  |  Ходы: %d  |  Пары: %d/%d",
			size, size, moves, pairsFound, totalPairs), bold))
		displayBoard(board, revealed, size)

		if board[r1][c1] == board[r2][c2] {
			fmt.Println(colorize("✅ Пара найдена!", green))
			pairsFound++
			time.Sleep(1 * time.Second)
		} else {
			fmt.Println(colorize("❌ Не совпало.", red))
			time.Sleep(time.Duration(timeout) * time.Second)
			if timeout == 0 {
				time.Sleep(1 * time.Second)
			}
			revealed[r1][c1] = false
			revealed[r2][c2] = false
		}
	}

	elapsed := int(time.Since(start).Seconds())
	fmt.Println(colorize(fmt.Sprintf("\n🎉 Поздравляем! Вы завершили игру за %d ходов и %d секунд.", moves, elapsed), green))

	best, ok := records[key]
	if !ok || moves < best.Moves || (moves == best.Moves && elapsed < best.Time) {
		records[key] = Record{Moves: moves, Time: elapsed, Date: time.Now().Format(time.RFC3339)}
		saveRecords(records)
		fmt.Println(colorize(fmt.Sprintf("🏆 Новый рекорд для размера %d×%d!", size, size), yellow))
	} else {
		fmt.Println(colorize(fmt.Sprintf("Лучший результат для этого размера: %d ходов за %d сек.", best.Moves, best.Time), blue))
	}
}
